### ISSUE TITLE
solve_cubic: batched gradients are NaN for three-real-root rows whenever the batch also holds a one-real-root row (D>0 branch computes sqrt(D) and -Q/AD on every |R|>0 row)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4334]
## Summary

`solve_cubic`'s batched backward is `nan` for every three-real-root row (`D < 0`, `R != 0`) as soon as the same batch also holds a one-real-root row (`D > 0`, `Q != 0`). Each row differentiates fine on its own; the contamination comes from the `D > 0` branch, which evaluates `sqrt(D)` and then `-Q / AD` over **every** row with `|R| > 1e-16`, not only the `D > 0` rows. `solve_quartic` inherits it through its resolvent cubic, so a batch mixing a four-real-root quartic with a two-real-root one gets `nan` gradients for the four-root row.

This is independent of #4290: it hits well-conditioned, distinct roots, and #4299 / #4329 (which fix the `acos` boundary inside the `D <= 0` branch) do not change it. Found while checking those two PRs against a random batch of repeated-root cubics: 26,282 of 50,000 rows had non-finite gradients on `main` and on both heads, and every one of them was finite when run alone.

## Reproduce

```python
import torch
from kornia.geometry.solvers import solve_cubic, solve_quartic

three = [1., -7., 14., -8.]   # (x-1)(x-2)(x-4): D < 0, three real roots, R != 0
one   = [1.,  0.,  1., -2.]   # (x-1)(x^2+x+2): D > 0, one real root, Q != 0

for rows in ([three], [three, one]):
    c = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    solve_cubic(c).sum().backward()
    print(len(rows), "row(s): grad of the three-root row =", c.grad[0].tolist())
# 1 row(s):  [-7.0, -1.0, -1.9e-16, -1.7e-16]
# 2 row(s):  [nan, nan, nan, nan]

q4 = [1., -10., 35., -50., 24.]   # (x-1)(x-2)(x-3)(x-4)
q2 = [1., 0., 0., 0., -16.]       # x^4 - 16: two real roots
for rows in ([q4], [q4, q2]):
    c = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    solve_quartic(c).sum().backward()
    print(len(rows), "row(s): grad of the four-root quartic =", c.grad[0].tolist())
# 1 row(s):  [-10.0, -1.0, 0.0, 0.0, 0.0]
# 2 row(s):  [nan, nan, nan, nan, nan]
```

Same in float32. The forward values are correct in every case; only the gradient is poisoned.

## Mechanism

`kornia/geometry/solvers/polynomial_solver.py`, `D > 0` branch on `main` (`ea07adac`):

```python
mask_R_positive = R_abs > 1e-16
if torch.any(mask_R_positive):
    AD[mask_R_positive] = torch.pow(R_abs[mask_R_positive] + torch.sqrt(D[mask_R_positive]), 1 / 3)
    ...
    BD[mask_R_positive] = -Q[mask_R_positive] / AD[mask_R_positive]
x0_D_positive = AD[mask_D_positive] + BD[mask_D_positive] - b_a_3[mask_D_positive]
```

`mask_R_positive` is a condition on `R`, not on `D`. For a `D < 0` row it computes `sqrt(D) = nan`, so `AD = nan` for that row. The row is never *read* (only `mask_D_positive` rows reach `x0_D_positive`), but `BD[...] = -Q / AD` is still a node in the graph, and `DivBackward0` for the unused row is `0 / nan = nan`, which lands on `Q.grad` and from there on every coefficient. `torch.autograd.detect_anomaly()` names exactly that node: `Function 'DivBackward0' returned nan values in its 0th output.`

The existing `test_gradcheck` fixture `(2x+1)(x-2)(x+3)` has `R == 0`, so it is excluded by `mask_R_positive` and never sees this; a single-row test cannot see it either.

## Suggested fix

Restrict the `D > 0` computations to `mask_D_positive & mask_R_positive` rows (index once, compute on the compacted rows, scatter back), so no `sqrt` of a negative `D` is ever built. A pin: a two-row batch of the two cubics above, `backward()`, assert finite — it fails on `main` and cannot pass by accident. Adding the quartic pair as well covers `solve_quartic`'s resolvent path.

Refs #4290, #4299, #4329, #4229 (the two `sqrt(clamp(...))` sites it lists in `solve_quartic` are on the same branch structure).

## Environment

kornia `main` at `ea07adac`, torch 2.14.0, CPU, macOS.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Comment by Jeevang1-epic** [comment]
Hi @ducha-aiki, I'd like to work on this one if it's available.I went through the reproduction and the branch interaction described here. I'll focus on restricting the D > 0 computation to the actual mask_D_positive rows so the negative-D rows never build the invalid sqrt/division autograd path, while preserving the existing forward results. I'll also add regression coverage for the mixed cubic batch and check the quartic resolvent case described in the issue.

If the scope looks right, could you assign this to me?
